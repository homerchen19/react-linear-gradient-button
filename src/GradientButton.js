import React from 'react';
import PropTypes from 'prop-types';
import styled, { css } from 'styled-components';
import { color, fontSize, borderRadius } from 'styled-system';
import { prop, ifProp } from 'styled-tools';

import { getLinearGradient, getPadding, filterProps } from './utils';

const GradientBackground = styled.button`
  position: relative;
  display: flex;
  box-sizing: border-box;
  padding: ${prop('borderWidth')}px;
  border: 0;
  outline: 0;
  background-image: linear-gradient(
    ${ifProp(
      ({ angle }) => angle.includes('deg'),
      prop('angle'),
      props => `to ${prop('angle')(props)}`
    )},
    ${props => getLinearGradient(props.theme, props.gradient)}
  );
  cursor: pointer;

  ${ifProp(
    'disabled',
    css`
      cursor: not-allowed;
      opacity: 0.65;
    `
  )};

  ${borderRadius};
  ${color};
  ${fontSize};

  text-decoration: none;
`;

GradientBackground.propTypes = {
  angle: PropTypes.string.isRequired,
  borderWidth: PropTypes.number.isRequired,
  disabled: PropTypes.bool.isRequired,
  gradient: PropTypes.arrayOf(PropTypes.string),
  theme: PropTypes.string.isRequired,
  ...borderRadius.propTypes,
  ...color.propTypes,
  ...fontSize.propTypes,
};

GradientBackground.defaultProps = {
  color: '#ae3560',
  fontSize: 16,
};

const Inner = styled.div`
  width: 100%;
  height: 100%;
  padding: ${props => getPadding(props.padding)};
  outline: 0;

  ${ifProp(
    { disabled: false },
    css`
      transition: ${prop('transition.property')} ${prop('transition.duration')}s
        ${prop('transition.timingFunction')} ${prop('transition.delay')}s;

      &:hover {
        background: transparent;
        color: #fff;
      }
    `
  )};

  ${borderRadius};
  ${color};
`;

Inner.propTypes = {
  disabled: PropTypes.bool.isRequired,
  padding: PropTypes.oneOfType([
    PropTypes.number,
    PropTypes.string,
    PropTypes.array,
  ]).isRequired,
  transition: PropTypes.objectOf(
    PropTypes.oneOfType([PropTypes.number, PropTypes.string])
  ).isRequired,
  ...borderRadius.propTypes,
};

const GradientButton = ({ children, ...props }) => (
  <GradientBackground
    {...filterProps({
      props,
      allowedProps: Object.keys(GradientBackground.propTypes),
      options: {
        withDOMProps: true,
        mapProps: {
          background: ({ value }) => ({ key: 'bg', value }),
        },
      },
    })}
  >
    <Inner
      {...filterProps({
        props,
        allowedProps: Object.keys(Inner.propTypes),
        options: {
          mapProps: {
            background: ({ value }) => ({ key: 'bg', value }),
            borderRadius: ({ key }) => ({
              key,
              value: props.borderRadius - (props.borderWidth + 1),
            }),
          },
        },
      })}
    >
      {children}
    </Inner>
  </GradientBackground>
);

GradientButton.propTypes = {
  angle: PropTypes.string,
  background: PropTypes.string,
  borderRadius: PropTypes.number,
  borderWidth: PropTypes.number,
  children: PropTypes.oneOfType([PropTypes.node, PropTypes.string]).isRequired,
  disabled: PropTypes.bool,
  gradient: PropTypes.arrayOf(PropTypes.string),
  padding: PropTypes.oneOfType([
    PropTypes.number,
    PropTypes.string,
    PropTypes.array,
  ]),
  theme: PropTypes.string,
  transition: PropTypes.objectOf(
    PropTypes.oneOfType([PropTypes.number, PropTypes.string])
  ),
  ...color.propTypes,
  ...fontSize.propTypes,
};

GradientButton.defaultProps = {
  angle: 'right',
  background: '#fff',
  borderRadius: 20,
  borderWidth: 2,
  disabled: false,
  gradient: null,
  padding: 10,
  theme: 'Vanusa',
  transition: {
    property: 'all',
    duration: 0.2,
    timingFunction: 'ease-in-out',
    delay: 0,
  },
};

export default GradientButton;
