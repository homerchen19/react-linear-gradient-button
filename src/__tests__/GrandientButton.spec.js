import React from 'react';
import TestRenderer from 'react-test-renderer';
import 'jest-styled-components';

import GradientButton from '../GradientButton';

describe('GradientButton', () => {
  it('should render correctly with default style', () => {
    const tree = TestRenderer.create(
      <GradientButton>test button</GradientButton>
    ).toJSON();

    expect(tree).toMatchSnapshot();
    expect(tree.type).toBe('button');
    expect(tree).toHaveStyleRule('padding', '2px');
    expect(tree).toHaveStyleRule('border-radius', '20px');
    expect(tree).toHaveStyleRule('color', '#ae3560');
    expect(tree).toHaveStyleRule('font-size', '16px');
    expect(tree).toHaveStyleRule(
      'background-image',
      'linear-gradient( to right,#DA4453 0%,#89216B 100% )'
    );
  });

  it('should render correctly with disabled props', () => {
    const tree = TestRenderer.create(
      <GradientButton disabled>test button</GradientButton>
    ).toJSON();

    expect(tree).toHaveStyleRule('cursor', 'not-allowed');
    expect(tree).toHaveStyleRule('opacity', '0.65');
  });

  it('should render correctly with fancy props', () => {
    const tree = TestRenderer.create(
      <GradientButton
        gradient={['#f00b47', '#0f6bb6']}
        angle="30deg"
        padding={[10, 20]}
        borderRadius={10}
        borderWith={3}
        color="#123456"
        fontSize={12}
      >
        test button
      </GradientButton>
    ).toJSON();

    expect(tree).toMatchSnapshot();
    expect(tree).toHaveStyleRule(
      'background-image',
      'linear-gradient( 30deg,#f00b47 0%,#0f6bb6 100% )'
    );
  });
});
