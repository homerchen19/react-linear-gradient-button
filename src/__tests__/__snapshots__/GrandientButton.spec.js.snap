// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GradientButton should render correctly with default style 1`] = `
.c0 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  padding: 2px;
  border: 0;
  outline: 0;
  background-image: linear-gradient( to right,#DA4453 0%,#89216B 100% );
  cursor: pointer;
  border-radius: 20px;
  color: #ae3560;
  background-color: #fff;
  font-size: 16px;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c1 {
  width: 100%;
  height: 100%;
  padding: 10px;
  outline: 0;
  -webkit-transition: all 0.2s ease-in-out 0s;
  transition: all 0.2s ease-in-out 0s;
  border-radius: 17px;
  background-color: #fff;
}

.c1:hover {
  background: transparent;
  color: #fff;
}

<button
  className="c0"
  color="#ae3560"
  disabled={false}
  fontSize={16}
>
  <div
    className="c1"
    disabled={false}
  >
    test button
  </div>
</button>
`;

exports[`GradientButton should render correctly with fancy props 1`] = `
.c0 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  padding: 3px;
  border: 0;
  outline: 0;
  background-image: linear-gradient( 30deg,#f00b47 0%,#0f6bb6 100% );
  cursor: pointer;
  border-radius: 10px;
  color: #123456;
  background-color: #fff;
  font-size: 12px;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c1 {
  width: 100%;
  height: 100%;
  padding: 10px 20px;
  outline: 0;
  -webkit-transition: all 0.2s ease-in-out 0s;
  transition: all 0.2s ease-in-out 0s;
  border-radius: 6px;
  color: #123456;
  background-color: #fff;
}

.c1:hover {
  background: transparent;
  color: #fff;
}

<button
  className="c0"
  color="#123456"
  disabled={false}
  fontSize={12}
>
  <div
    className="c1"
    color="#123456"
    disabled={false}
  >
    test button
  </div>
</button>
`;
